src/Swarm/Util/WindowedCounter.hs

```diff
@@ -0,0 +1,135 @@
+-- {ticks for which that state existed}
+-- to the
+-- {total number of ticks spanned by the window}.
+--
+-- The primary use case is in displaying the "activity level" of a robot.
+--
+-- == Efficiency considerations
+--
+-- The data retention of the window shall be maintained externally by
+-- invoking the 'discardGarbage' function. However, we should not
+-- unconditionally invoke this function upon each game tick.
+--
+-- For efficiency, we do not want to iterate over every robot
+-- upon every tick; we only want to "visit" a robot if it is actually
+-- doing work that tick.
+-- Because of this, there may be some ticks in which the oldest element
+-- that is still stored falls outside of the nominal retention window.
+--
+-- One might think we could perform garbage collection whenever we execute queries.
+-- However, in the context in which the view powered by the query is generated, we
+-- are not permitted to mutate the "state" of the game
+-- (the type signature of the rendering function is @AppState -> Widget Name@).
+--
+-- Therefore, when we perform "queries" on the window, we must apply some
+-- filtering to exclude the "stragglers"; data members that have already fallen outside
+-- the window but have not yet been "garbage collected".
+-- We use a 'Set' to allow this filtering to be performed in @O(log n)@ time.
+--
+-- In the worst case, the entire dataset may "age out" without being garbage collected,
+-- so that an @O(log n)@ filtering operation might be performed upon every "frame refresh"
+-- of the UI view.
+-- However, we also store the largest element of the window separately from the 'Set' so that
+-- we can compare against it for a @O(1)@ short-circuited path once every member ages out.
+data WindowedCounter a = WindowedCounter
+  { _members :: Set a
+  , _lastLargest :: Maybe a
+  -- ^ NOTE: It is possible that '_lastLargest' may not exist in the 'Set'.
+  , _nominalSpan :: a
+  -- ^ Data retention window
+  }
+  -- NOTE: deriving 'FromJSON' circumvents the protection offered by "smart constructors",
+  -- and the 'ToJSON' instance may expose internal details.
+  deriving (Eq, Show, Generic, FromJSON, ToJSON)
+
+mkWindow ::
+  -- | Window span
+  a ->
+  WindowedCounter a
+mkWindow = WindowedCounter Set.empty Nothing
+
+-- | Return the ratio of {members in the window} to the {integral span
+-- represented by the window} (actually, the beginning of the window up to
+-- some reference tick).
+--
+-- The "reference tick" must be at least as large as the largest
+-- element of the window.
+--
+-- A fully-contiguous collection of ticks would have an occupancy ratio of @1@.
+getOccupancy ::
+  Integral a =>
+  -- | reference tick
+  a ->
+  WindowedCounter a ->
+  Ratio a
+getOccupancy referenceTick (WindowedCounter s lastLargest nominalSpan) =
+  if Set.null s || maybe False (< referenceTick - nominalSpan) lastLargest
+    then 0
+    else case Set.lookupMin s of
+      Nothing -> 0
```

Comment:
TODO: I think I forgot to implement the filtering step described in the long Haddock. Should probably use `Set.split` here.